webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/domain/reports/ComplianceLevel.scala

```diff
@@ -111,57 +110,111 @@ object CompliancePercent {
   }
 
 
+  // hardcoding the precision levels
+  val divisers: Seq[Long] = Seq(1, 10, 100, 1000, 10000, 100000)
+  val hundreds: Seq[Long] = Seq(100, 1000, 10000, 100000, 1000000, 10000000)
+
   /*
-   * Ensure that the sum is 100% and that no case disapear because it is
-   * less that 0.01% (assuming precision 2).
-   *
-   * Rules are:
-   * - always keeps at least 10e(-precision)% for any case
+   * Computes the compliance with precision from 0 to 5 (max)
+   * It enforces that:
+   * * always keeps at least 10e(-precision)% for any case
    *   (given that we have 14 categories, it means that at worst, if 13 are at the mean in place of ~0%;
-   *   the last one will be false by 13*10e(-precision)%. For precision = 0, it can reach 13% in a pathological case.
-   * - the biggest value is rounded to compensate
-   * - in case of equality (ex: 1/3 success, 1/3 NA, 1/3 error), the biggest is
-   *   chosen accordingly to "ReportType.level" logic.
-   *
+   * * the sum sums to 100, by having the largest one computed with 100 - sum of the others
+   * * best effort to keep the ordering in percentage
+   *   it does that by adding the extra 10e(-precision)% to the total, and computing the percentage
+   *   by dividing the remaining percent with this
+   * All computation is in Long by multiplying the values with the power of ten
+   * and then truncated by converting to double and dividing after by the power of ten
    */
   def fromLevels(c: ComplianceLevel, precision: Int): CompliancePercent = {
-    // the value for the minimum percent reported even if less (see class description)
-    val MIN_PC = BigDecimal(1) * BigDecimal(10).pow(- precision) // since we keep precision digits in percents
-
-    if(c.total == 0) {  // special case: let it be 0
+    val total = c.total
+    if (total == 0) { // special case: let it be 0
       CompliancePercent()(precision)
     } else {
-      val total = c.total // not zero
-      def pc_for(i:Int) : Double = {
-        if(i == 0) {
+      // these depends on the precision
+      val diviser = divisers(precision)
```

Comment:
@fanf do you know what's the cost of doing an assert ? I'll need to assert that prevision is <=5

Reply:
create a case class precision with case object Level1 to Level5 and its value setted. It should not be perf impacting (object are global, and accessing a static field is very quick, and the traduction to level is only done at border like API so not important)